src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,788 @@
+                                {
+                                    this.RenewLease(partition);
+                                    renewedLease = true;
+                                }
+                            }
+                            else
+                            {
+                                this.DrainLease(partition, CloseReason.Shutdown);
+                                this.RenewLease(partition);
+                                renewedLease = true;
+                                drainedLease = true;
+                            }
+                        }
+                    }
+
+
+                }
+                await this.BalanceLeases(partitionDistribution, partitions, ownershipLeaseCount, response);
+
+                // If shutdown is requested and the worker releases all ownership leases, then set the response.IsReleasesAllLease to true to notify the partitionManagerLoop to stop.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    response.IsReleasedAllLease = (ownershipLeaseCount == 0);
+                }
+                return response;
+            }
+
+            // This method performs a worker-level partition rebalancing process.
+            // The idea is to calculate the expected number of leases per worker in a fully balanced scenario
+            //  (a.k.a their partition quota) and for each worker to steal partitions from others
+            // until they have met their quota. 
+            // A few remarks:
+            // (1) The quota of partitions per worker is the number of partitions divided by the number of workers. If these two number can not be evenly divided, then the difference between quota of partitions per workers should not exceed one.
+            // (2) Workers only steal from workers that have exceeded their quota
+            // Exception will be thrown if the update operation fails due to outdated etagso that worker could re-read the table again to get the latest information.
+            // Any other exceptions will be caught to logs. 
+            async Task BalanceLeases(Dictionary<string, List<TableLease>> partitionDistribution, Pageable<TableLease> partitions, int ownershipLeaseCount, ReadTableReponse response)
+            {
+                // Check if there is any other active worker, if not, skip the balance process.
+                if (partitionDistribution.Count == 0)
+                {
+                    return;
+                }
+
+                int averageLeasesCount = (partitions.Count()) / (partitionDistribution.Count + 1);
+                if (averageLeasesCount < ownershipLeaseCount)
+                {
+                    // Already have enough leases, and thus no need to steal, just stop the balancing.
+                    return;
+                }
+
+                //Check all the other active workers when my ownership lease number is not larger than the average number.
```

Comment:
nit: let's add a space between the "//" and the first character
Also suggesting a small change in phrasing
```suggestion
                // If this worker does not own enough partitions, search for leases to steal
```